crates/wasmtime/src/store.rs

```diff
@@ -414,6 +465,23 @@ unsafe impl TrapInfo for Store {
     fn max_wasm_stack(&self) -> usize {
         self.engine().config().max_wasm_stack
     }
+
+    fn out_of_gas(&self) {
+        #[derive(Debug)]
+        struct OutOfGas;
+
+        impl fmt::Display for OutOfGas {
+            fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+                f.write_str("all fuel consumed by WebAssembly")
+            }
+        }
+
+        impl std::error::Error for OutOfGas {}
+
+        unsafe {
+            wasmtime_runtime::raise_lib_trap(wasmtime_runtime::Trap::User(Box::new(OutOfGas)))
```

Comment:
Can you say more about the expected way to run this in an executor-loop system with timeslicing (as in Lucet's fuel implementation)?

Specifically, is the common case that we'll run for a timeslice and then unwind here using the trap mechanism back to the Wasm entry point, at which point some higher-level wrapper might yield a future or similar?

My two concerns are:

- Efficiency: invoking the unwind mechanism and heap-allocating an error is somewhat heavyweight;
- Ability to resume: it looks like `raise_lib_trap` eventually invokes `Unwind` in wasmtime-runtime's helpers.c, which uses longjmp to escape the Wasm stack frames. Does this mean that we can't resume at the point where fuel was exhausted (i.e., it's a terminating trap rather than a resumable one)?

All of my concerns above go away if we have a way to plug in a custom out-of-gas handler; I couldn't find an API that would let one do this (one would need to pass in a custom `TrapInfo` trait impl I think?) though I may have missed it. Or alternately, is the more complete interface coming with later async work?

Reply:
Oh sure yeah, the "raise a trap" implementation here is primarily just so we can test it in this PR without all the async bits landed yet. Additionally it helps us fuzz the implementation well by ensuring wasm is always consuming fuel when executing. For an async yield-every-so-often implementation my plan is to:

* Still have this bake into a `Store` without the ability to register a custom handler (although that's still possible, but fraught with correctness issues on the caller's part)
* Instead of raising a trap here this would simply switch off the fiber. With async we are guaranteed that all wasm is always executing on a fiber, so this is possible.
* The `Store` would have some sort of flag/configuration where in async mode you could request that N fuel is used up and when finished it yields to the current future and then injects N more fuel when it comes back.

The implementation would be relatively simple, basically [calling suspend](https://github.com/bytecodealliance/wasmtime/blob/64ab442767789571b69a49caacf6aae2967664f6/crates/wasmtime/src/store.rs#L532-L534) here aftter notifying ourselves to the future executor saying we're already ready for another `poll`. I would imagine that this would all be guarded by a basic `if` in this out-of-gas handler which either traps or yields.

Efficiency-wise I think it should be quite fast because it's a fiber switch and no unwinding happens (not even longjmp). Upon resumption we'd simply return from this function and wasm would keep going. Resumption-wise we should be good as well due to fibers and whatnot. Basically the trap stuff won't happen at for the timeslicing, it's just a way for me to land this PR before the async fiddly bits are here.

Reply:
Makes sense! Happy to see this go in as-is, then.